### DISCUSSION TITLE
Adding auctions to PBN files

### DISCUSSION THREAD

**Discussion by Rick-Wilson** [ADavidBailey/Practice-Bidding-Scenarios#40] [Ideas]
I'd like to have auctions in the PBN files.  This will give a good source of hands for bidding practice tools.

What I need are hands for each scenario where the auction is included in the PBN files.  @ADavidBailey has started working on this on a one-by-one basis.  I have put together a script to run through all the /pbn files to create Auction versions.  The script runs BBA with command line parameters, then waits 15 seconds, and kills BBA.  Then on to the next file.

This works mostly okay, but I'm having a lot of problems with BBA command line parameters - for example, the archive folder isn't working, and the CC1 and CC2 options aren't working.  I have workarounds for all of these, but it will be a lot simpler when the new version of BBA is released.

One issue I have is that lots of the hands in the PBN files don't actually use the scenario they are designed for.  I am thinking of writing a Bridge Composer script to scan the PBNs and filter out hands where the BBA auction doesn't make use of the intended scenario.  This could also give a hit % for each hand, giving an idea of how well the script and BBA are in agreement.

It won't always be easy to tell whether it's a hit or not.  For some conventions, BBA adds an auction note, which could be used for a filter.  For others, it doesn't, and a more complex search will be required.  I'm thinking that we might be able to have a regular expression string for each scenario, which can be run in the filter to decide if it worked or not, on a hand-by-hand basis.

As far as using these files goes, I have written a Bridge Composer script which reads a PBN file with auction, and generates an HTML file, formatting for bidding practice.  The HTML file is converted to a PDF, then printed two-sided.  This way, in face-to-face lessons, people can bid hands just from the printouts, then turn the page over to see how the auction should have gone.
[Jacoby and Texas Transfers x5 Bidding Sheets.pdf](https://github.com/user-attachments/files/17177148/Jacoby.and.Texas.Transfers.x5.Bidding.Sheets.pdf)

**Comment by Rick-Wilson** [comment]
Here's the command procedure to run BBA on one file, with the 15 second kill.
[bba_one_file.cmd.zip](https://github.com/user-attachments/files/17177151/bba_one_file.cmd.zip)

**Reply by ADavidBailey** [reply]
I tried using this.  It doesn't suit my environment.  I have a makeBBA.py that creates a makeBBA.cmd file that works.
https://github.com/ADavidBailey/Practice-Bidding-Scenarios/blob/main/makeBBA.cmd  (truncated for testing)

There are some issues with the BBA command line processing.  I'm waiting for Edward to fix it.

BBA reads a file from the pbn folder and archives it to the bba folder with the same name -- without the dealer tagged to the end of the name.  (All of the files in pbn have a -N, -E, -S, or -W appended to the name.  I need this to do the rotation.)

So, the only issue is that I have to manually kill the UI.

I know, my environment sucks; but, it's working for me.

**Comment by Rick-Wilson** [comment]
Here's the powershell script to run the BBA script for all PBNs.

This necessarily takes 15 seconds per hand, so 188 hands takes about 45 minutes to run.

[bba_all_pbns.ps1.zip](https://github.com/user-attachments/files/17177152/bba_all_pbns.ps1.zip)

**Reply by ADavidBailey** [reply]
@EdwardPiwowar just posted the new release of BBA.   It has --AUTOCLOSE.

**Comment by Rick-Wilson** [comment]
Here's another example of the Bidding Sheets, this one from a /pbn script file run through BBA (500 hands).
[Uploading 1N_with_Singleton-S_Archive Bidding Sheets.HTML.pdf…]()

**Reply by ADavidBailey** [reply]
Is this link doing what you intend?  For me, it just re-opens this Discussion.  [Here's the link](https://github.com/ADavidBailey/Practice-Bidding-Scenarios/discussions/40)

**Comment by Rick-Wilson** [comment]
One thing I need for this is N/S rotations of the hands.  I can do this with a Bridge Composer script, which will work post-auction, or with a modification of David's Python (?) rotation script.

**Comment by ADavidBailey** [comment]
You've been busy!  i, too, am looking forward to the next release of BBA.

**Comment by ADavidBailey** [comment]
I just got around to looking at this discussion, today.  Sorry 'bout that.

**Comment by ADavidBailey** [comment]
I like having everything in the pbn file.  I also like having a 1-line summary of BBA and DDS information.  That's a lot to squeeze into a single line.  

These single lines could be easily extracted from the pbn and put into a csv file for easy reading/searching.
@ThorvaldAagaard

**Reply by ADavidBailey** [reply]
Hmmm, can we use regular expressions in a csv file?  I like the readability of the files in my bba+ folder.
https://github.com/ADavidBailey/Practice-Bidding-Scenarios/tree/main/bba%2B

And, I just discovered that VSCode supports regular expressions.

**Comment by Rick-Wilson** [comment]
What I'm looking for here is a way to search for hands that were generated which match the scenario.   Let's look at two examples from your /BBA+ folder:

4th suit forcing - when I look at your Fourth_Suit_Forcing.pbn.txt file, and I copy the 500 PBN lines into a text editor, then I search for "Fourth suit", I get 379 matches.  So this is 76% of the lines.  So I'm thinking your original script is 76% successful in generating 4th suit forcing deals.  And if I want to generate a PBN file with just those 379 hands that use FSF, I could easily write a script to search for "Fourth suit" and just keep those hands.  (It's a little harder when searching the PBN files themselves, because they are multi-line, but still we have tools that will deal with that.)

Now let's look at your Game_Overcalls.pbn.txt file.  There's not a simple text phrase to search for here, to see if we get a game-level overcall or not.  There is no bidding note saying anything about the overcall.  Looking at the first few lines we can see:

```
brd# contract score  par     north            east             south            west             | auction...   | notes
---- -------- -----  ------  ---------------- ---------------- ---------------- ---------------- | ------------------------- 
   1  5HX-S     650          32.87.A542.QJ942 QJ64.AQ5.QJ983.7 A7.KJT96432.7.A5 KT985..KT6.KT863 | 1D-1H-1S-P-2S-3H-3S-4H-P-P-4S-P-P-5H-X
   2  4H-S     -100          K853..T73.AQT652 J.AQ72.AJ642.974 QT9.KJT98653.K.J A7642.4.Q985.K83 | 1D-4H
   3  4H-S      480          KQJ642.AQ.A84.T2 3.9.KT63.AKQ9753 A5.KJT87632.J7.4 T987.54.Q952.J86 | 1C-4H
```

The first hand overcalled 1H instead of 4H.  So that's a "miss", and the next two are "hits".  To find the hits here, we can try a regular expression.  This expression works: " 1.-4." (without the quotes).  This finds 412 occurrences.  This gives us an 82% hit rate.  And if I want to find just the hands where a game-level overcall occurred, I could to a search of the PBN using that regular expression.

So what I was thinking is, maybe we should have a regular expression per scenario, which could be used to find the "hits".   We could probably add a script which would search all the BBA files using the scenario-specific regular expression, and produce a table showing the hit % for each scenario.  We could also have a script which would filter the BBA output down to just the hit rows.

Also note: the regular expression I wrote works on your DDS output, not the PBN (the auction is formatted differently in the PBN).  But that just means the regular expression would need to be slightly different to run on the actual PBN file.

**Reply by ADavidBailey** [reply]
There are a few 1.-5 files.

If you'll write the 'regular expressions', I'll add them to the scenarios as a comment.  The 1.-4 and 1.-5 are the most complicated RE's I've ever created. :)

I've played with [Python RegEx](https://www.w3schools.com/python/python_regex.asp), it's really helpful.

**Reply by Rick-Wilson** [reply]
I changed that one to ^1.\s+[4,5]. and it finds 431 matches, so 86%.

I have a Bridge Composer script now which will apply the filters, but we still need to generate the regular expressions for each scenario.

**Reply by ADavidBailey** [reply]
We should create a regex folder with file names corresponding to those in my bba folder.   Then we could create a program that would read the files in regex and apply it to the corresponding bba file and put the results in a bba-filtered folder.

**Comment by Rick-Wilson** [comment]
(In a regular expression, the dot "." matches any single character).

**Comment by Rick-Wilson** [comment]
So what I'm thinking is you could have a series of files, starting with your scenario button, the rest script-generated, looking something like this:

scenario-button
   -> scenario-text (no button info)
   -> pbn file (hands only)
   -> bba file (hands and auctions)
   -> double dummy (maybe from Bridge Composer, which can be run from the command line)
   -> bba filtered file (hits only)
   -> rotations (via Bridge Composer)
   -> N/S bidding sheets (and eventually other similar tools)

This would really be a tree structure, because you can have multiple children from any of these levels, with various statistics/metrics along with the files.

**Reply by ADavidBailey** [reply]
| Program | from | to |
| ------------------ | ------------------------------ | ------------------------------------ |
| extract.py   |  PBN | dlr |
| makePBN.py | dlr | makePBN.cmd
| makePBN.cmd | dlr | pbn |
| commentStats.py | pbn | pbn |
| rotate.py | pbn | pbn-rotated-for-4-players |
| PBNtoLIN.py. | pbn-rotated-for-4-players | lin-rotated-for-4-players |
| makeBBA.py | pbn | makeBBA.cmd |
| makeBBA.cmd | pbn | bba |
| bbaSummary.py | bba | bba-summary |
| ???????????  | bba-summary | bba-filtered

**Comment by ADavidBailey** [comment]
I've created summaries of all of the files in bba folder.  They're in the bba+ folder.

**Comment by ThorvaldAagaard** [comment]
From training of BEN perspective it is very interesting to have only deals, that we know is bid according to agreement.

Just to give you an update of how training material is produced you can git my list.

- I have collected about 1 mill interesting deals
- I let BBA bid the 1 mill hands according to system, and let BBA's Single Dummy decide the score
- I calculate DD-score using Bridge Composer
- I remove all deals, where the actual score is more than 2000 (configurable) different than the PAR-score
- I remove all deals that was doubled and th contract was making

Finally I have a limit on number of times I want to see a specific bidding sequence - actually 50

So what I will do when the quality of the files you create are good I will probably add those as a starting point, and then add the above files I produce, and thus skipping deals where I already have enough for a given sequence.

**Reply by ADavidBailey** [reply]
I'm confused about how I should run BBA.  Is there 1 way to run it that is useful to all?

1. Which bots should I use?  Should I create files with multiple bots?
2. Which convention cards should I use?  Should I create files multiple CCs?
3. SD vs DD?  What should I use?  Why?

From the point of view of my scenarios, my primary concern is how they work on BBO with the BBO Robots.

From the point of view of Ben...?

From the point of view of...?

**Reply by ThorvaldAagaard** [reply]
I have a special CC for training BEN for playing with GIB, but in an ideal world each set of deals is just one convention, so I can pick the sets I want.

I the real world it will not work as slambidding, RKCB-0314 or RKCP-1430 cubbies etc might be different, so I will probably have to rebid all the hands anyway, but I will still have better hands for the different agreements.

**Reply by ThorvaldAagaard** [reply]
But I think you should focus on creating deals for real players, and not be so concerned about BEN, it was just to give you an idea of how I am creating training data.

**Comment by Rick-Wilson** [comment]
You need to add Gambling to the BBA convention card.

**Reply by ThorvaldAagaard** [reply]
It is important that when bidding a set, the correct convention is selected, otherwise stats will not be calculate able

**Reply by ADavidBailey** [reply]
I did so.  And, I rebuilt the dependant files pbn, rotated, pba...

**Comment by ADavidBailey** [comment]
This mornin' I woke up thinkin' 'bout rotation.

When the pbn file is created, the [Dealer ...] record,  [Board Number ...] record, and [Deal...] record agree with one another.  However, the [Board Number...] does NOT correspond to its position in the file -- the Record Number.  Rotation merely forces all others to correspond to the Record Number.

The initial pbn is simple.  Only the Board Number, the Dealer, and the Deal need to be 'rotated'.  BBA adds [North...], [East...], [South...], [West...],  [Declarer...], [Score...]. [Auction...], and [Play...].

(Hmmm, what if you had [Dealer "W"] and [Auction 'S"]?  Declaring the same thing twice is poor design!  Why does [Score...] have a value of "NS"?  It has that value ALWAYS!!  You rotate the Score by toggling +/-.)

You can filter records from any pbn file and just re-sync the board number with the record's position in the file.

The 180 rotation is just a filter -- toss the even-numbered boards.  (would you mind playing only odd-numbered boards?)

I didn't mention [Vulnerable...].  I ain't goin' there.

I'll re-read this when I wake up.

The dealer suffix that I added to the pbn files convenient; but, redundant.  It's in all of the pbn [Dealer...] records.

**Reply by ADavidBailey** [reply]
Rotation should be done as a Bridge Composer script.  @Rick-Wilson, can your 'filter' scripts call a 'renumber' script?

Many that went through a bunch of deals in BC, deleting some, would want to renumber them before saving.

What does the Renumber Boards script do?  I tried Edit/Select All, Board/Renumber (Dealer was South, so I started with 3).  The result looks good.  EVERYONE CAN COMMENT ON THIS SPREADSHEET.

Here's the before/after:  https://docs.google.com/spreadsheets/d/e/2PACX-1vTb5_zc9Cb40AP-wUM8OEgY-eLcEUNYLUjTJdKHmxi0kAcLlnCDfy6XxHdreHeegZQM38Zm9QYI1Den/pubhtml

**Reply by ThorvaldAagaard** [reply]
Should vulnerability follow the number?
I did a renumber deal in Bridge Composer and it was very slow (But a very large file), so I created a script 
https://github.com/ThorvaldAagaard/Bridge-Robot-Utilities/blob/main/src/RenumberPBNBoards.py

**Reply by ADavidBailey** [reply]
None of my PBS Dealer code considers Vunerability.  I have no idea how I would/could do that.  Bridge Composer did not seem slow to me.  Much, Much less time that running BBA.  @Rick-Wilson mentioned that Bridge Composer could be run via a CMD file (I think).

**Reply by ThorvaldAagaard** [reply]
I haven't tried Bridge Composer from the command prompt, but according to documentation Rick is right

BridgeComposer also supports classic ".bat-file" scripting. See BridgeComposer "Help→Help Topics", topic "Command Line Interface".

**Reply by ADavidBailey** [reply]
@ThorvaldAagaard does your script to RenumberPBNBoards rotate Dealer, Deal, Declarer, Auction, and Play?

Did you look at my before/after spreadsheed, above?

**Reply by ThorvaldAagaard** [reply]
My script just read all boards and change the number from the beginning and save the file. Nothing else should be changed.

I looked at the before and after and noticed, that the bidding was different, but not more.

**Reply by ADavidBailey** [reply]
It seems like we've had this conversation before.  :)

If you want the intended auction, my deals require that the Dealer, Deal, Declarer are rotated.  And, if you want to rotate after the Auction and Play, you need to rotate those, too.

**Reply by ADavidBailey** [reply]
My 'rotate' is no longer dependant on the pbn suffix.  I will be removing those in the near future.
